### PR TITLE
Copy and convert tests form Mono

### PR DESCRIPTION
**Pull request by anthonylangsworth**
For issue #6

### CODE REVIEW

**Comment by tintoy** [comment]
Someone's been busy :)

**Comment by anthonylangsworth** [comment]
@tintoy Still have 84 failing tests to go. A bit of a slog at the moment.

**Comment by tintoy** [comment]
You're a machine!

Would it help you if we split up and farmed out some of the work? I think there may be one or two others keen to pitch in :wink:

**Comment by anthonylangsworth** [comment]
Thanks, @tintoy . I have added a note to let others know help is appreciated.

**Comment by peterwurzinger** [comment]
Hi @anthonylangsworth !
I fixed 2 - 3 tests, didn't have time for more :-(

But I analyzed a strange phenomenon in more depth:
When running the tests I faced the exception "Could not create the XML transformation identified by the URI ..." very often. Did you experience the same?

The strange thing about this is, that there is a class `System.Security.Cryptographic.CryptoConfig`, backed by some kind of static `HashMap<string,Type>`, and it actually **does** contain entries with the keys he throws the exception on. Is this a problem with my machine or is it a problem you also do face?

Cheers
Peter

**Comment by tintoy** [comment]
Just out of curiosity, @peterwurzinger, does the problem go away if you add `-parallel none` when running `dotnet test`?

**Comment by peterwurzinger** [comment]
Well I already had some issues with the parallelism of xUnit in other projects, the hint is really good. I'll try tomorrow morning, gonna go to sleep now - time offset makes life complicated sometimes ^^

**Comment by anthonylangsworth** [comment]
@peterwurzinger Thanks for the fixes.

To prevent the transformation errors, you'll need to run `build.cmd` again first. I made a change to the `CryptoConfig` class in `System.Security.Cryptography` to add the transformations exposed in the `System.Security.Cryptography.Xml` namespace.

**Comment by peterwurzinger** [comment]
Aaah, I see. Yep, it's working now, thanks a lot @anthonylangsworth . I'm going to carry on and fix some tests this afternoon - or "Early Tuesday morning" as you would call it ^^

**Comment by anthonylangsworth** [comment]
I'm glad you got it working, @peterwurzinger . We all just need to remember to `git pull` regularly. I'm going to try to get some more work done tonight, too.

**Comment by anthonylangsworth** [comment]
A lot of the remaining tests require implementing `FromXmlString` and `ToXmlString` on the `DSA` and `RSA` classes. Unfortunately, these are defined in `System.Security.Cryptography.dll` and not `System.Security.Cryptography.Xml.dll`. This is more complex than the `CryptoConfig` change and requires tests.

Unfortunately, I cannot get the `System.Security.Cryptography.Algorithm` project (part of `System.Security.Cryptography.dll`) to compile in Visual Studio. Merging from trunk makes the problem worse.

Therefore, we have a few options:
1. Someone else has a look at getting that project to compile. Perhaps I have missed something simple. This could be us or Microsoft.
2. We ask the Microsoft guys for the source code to these methods and their tests. Writing the XML parsing code is not hard, just time consuming. However, dropping in code and tests means we can just build it using `build.cmd` instead of Visual Studio, which works.

@tintoy and @peterwurzinger What are your thoughts?

**Comment by peterwurzinger** [comment]
Well, that's some kind of showstopper imho.
I guess a first shot in the dark would be porting those methods from the source of the current .NET-Framework (https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/rsa.cs and https://referencesource.microsoft.com/#mscorlib/system/security/cryptography/dsa.cs) to .NET-Core. But to be fair, I have absolutely no clue if they
a) are portable to .NET-Core at all and
b) will work with the Mono-tests.

Also I do not know where to get test-code from, but those methods absolutely need some! No way getting around that.

Unfortunately I'm not able to do much about the building-stuff, I'm not that deep into hacking .csproj-files. This has to be the first step imho, implementing methods and compile them with `build.cmd` would be quite annyoing.

**Comment by tintoy** [comment]
IMO, let's not stuff about - we've already done more than enough work to make it worth getting them involved :)

@steveharter @bartonjs do either of you have an opinion on how to proceed?

tl; dr; further testing will require porting of missing features in `System.Security.Cryptography.Algorithms` (`RSA.FromXmlString` and friends).

**Comment by bartonjs** [comment]
To/FromXmlString can't really work in .NET Core with the current assembly factoring.  They really shouldn't have ever existed in the first place, since they are only there as a part of Signed/Encrypted Xml.

So, rather than making those methods work in S.S.C.Algorithms, make some internal static methods in S.S.C.Xml and call them.

**Comment by anthonylangsworth** [comment]
OK. I'll see what I can do with the `RSAKeyValue` and `DSAKeyValue` versions of `FromXmlString` and `ToXmlString`. I originally thought I'd need access to private members of `RSA` and `DSA` but, after writing my previous comment, I think there may be another way.

**Comment by peterwurzinger** [comment]
So this sounds like we got a plan of how to proceed with getting the monotests running. Do you need help with those To/FromXmlString - Methods @anthonylangsworth ?

**Comment by anthonylangsworth** [comment]
Added `RSAKeyValue.LoadXml` and associated tests but the failure count did not drop as much as expected (down to 45). I'll hit `DSAKeyValue.LoadXml` tomorrow then see where we are at. After that it is slogging through the transformations.

Code reviews on the last few commits are welcome, too.

**Comment by peterwurzinger** [comment]
Looks good to me at first sight on my smartphone, I'll review it tomorrow in greater detail. Strange, that the fail counter didn't decrease to a lower number though.

**Comment by anthonylangsworth** [comment]
@tintoy Using an `XmlNamespaceManager` gives the following:

```C#
const string xmlDsigNamespacePrefix = "dsig";
XmlNamespaceManager xmlNamespaceManager = new XmlNamespaceManager(value.OwnerDocument.NameTable);
xmlNamespaceManager.AddNamespace(xmlDsigNamespacePrefix, SignedXml.XmlDsigNamespaceUrl);

XmlNode rsaKeyValueElement = value.SelectSingleNode(
    $"{xmlDsigNamespacePrefix}:{RSAKeyValueElementName}", xmlNamespaceManager);
if (rsaKeyValueElement == null)
{
    throw new CryptographicException(
        $"{KeyValueElementName} must contain child element {RSAKeyValueElementName}");
}

try
{
    Key.ImportParameters(new RSAParameters
    {
        Modulus = Convert.FromBase64String(
            rsaKeyValueElement.SelectSingleNode(
                $"{xmlDsigNamespacePrefix}:{ModulusElementName}", xmlNamespaceManager).InnerText),
        Exponent = Convert.FromBase64String(
            rsaKeyValueElement.SelectSingleNode(
                $"{xmlDsigNamespacePrefix}:{ExponentElementName}", xmlNamespaceManager).InnerText)
    });
}
catch (Exception ex)
{
    throw new CryptographicException(
        $"An error occurred parsing the {ModulusElementName} and {ExponentElementName} elements",
        ex);
}
```

The code is a few lines longer but the XPaths are easier to read. This is probably a better solution if you are dealing with multiple namespaces but you have to be careful of reusing/overwriting prefixes. Unsure which solution is categorically better.

**Comment by tintoy** [comment]
Do you really need to use interpolation for the prefix? It'd look a lot cleaner without it, and since it's basically an inline constant where the namespace manager is declared, there's not as much value to using it. I'm fairly rusty on the legacy XML / XPath API (a.k.a. `XmlDocument` and friends), but from memory it was fairly common to see code along the lines of:

```csharp
XmlNamespaceManager nsgmr = new XmlNamespaceManager();
nsmgr.Add("my", myNamespace);

node.SelectSingleNode("my:foo", nsmgr);
```

IIRC, the prefix can be anything as long as it matches the one passed into the namespace manager (i.e. no need to use constants).

**Comment by tintoy** [comment]
(am trying not to be irritatingly nit-picky, but am also aware that I probably am nevertheless being exactly that; sorry)

:wink:

**Comment by anthonylangsworth** [comment]
Your code is basically correct, although you need to reference the `XmlNameTable` from the document to consider existing namespace declarations. I use constants because I have been dragged up that way. I think we are both bike shedding now. In the absence of material differences, I'll move on.

**Comment by tintoy** [comment]
Sorry, yes, you're right on both counts :)

**Comment by peterwurzinger** [comment]
I messed up hard while pushing my changes for DsaKeyValue.LoadXml() ...
Can someone please help me out @tintoy  @anthonylangsworth ? I figured, that opening a [PR](https://github.com/anthonylangsworth/corefx/pull/1) in @anthonylangsworth 's repo felt like the cleanest thing to do to clean up the mess...
Sorry for that!

**Comment by tintoy** [comment]
@peterwurzinger all good, I can't merge that PR for you (don't have permissions to @anthonylangsworth repo), but your approach seems fine :-)

**Comment by anthonylangsworth** [comment]
Merged. Thanks, @peterwurzinger .  Yes, I had started that class but you got further than I did.

**Comment by peterwurzinger** [comment]
Just fixed a bug in my own code - not a great feeling but I guess it actually does work now ^^

EDIT: Hit the wrong button and accidentally closed the PR... Got to take care of my stubby fingers...

**Comment by Jaedson33** [comment]
So... Are there any failing tests now?

**Comment by tintoy** [comment]
Have you tried pulling down the latest code and running them? That'll probably be a more definitive answer than what any of us claim to remember :)

**Comment by Jaedson33** [comment]
I am very busy now, then it is easier to ask this to you.

**Comment by tintoy** [comment]
Fair enough. I think we're all pretty busy at the moment, so the best answer I can give for now is yes some tests are still failing. Sorry but it'll be a while before you're able to use this unless you're willing to take a copy of the source code.

**Comment by tintoy** [comment]
(taking a copy of the source code is a totally valid option, by the way, as long as you understand the risks involved)

**Comment by anthonylangsworth** [comment]
@Jaedson33 As of commit ac6dd84cfa1146d4ad47b52e6247773d130c4a20, there are still 49 tests failing.

**Comment by peterwurzinger** [comment]
In my opinion @Jaedson33 a production ready version of the `System.Security.Cryptography.Xml` - package will earliest be available with the release of netstandard2 and the next framework generation/version. Microsoft has (afaik) not announced any release dates yet, so imho this won't happen before summer '17.

If you're planning to write PoC-/academic code, you can clone the sources and build your own version of the library. Keep in mind, that some parts are still missing.
Cheers

EDIT: STUBBY FINGERS CLOSED THE PR AGAIN, AAARGH!

**Comment by Jaedson33** [comment]
I'll wait until all errors are corrected, then I will see what I do. 😉

**Comment by peterwurzinger** [comment]
I guess we need to discuss further steps in relation to [work of others](https://github.com/dotnet/corefx/pull/16545). Is fixing the failing tests still the way to go? Should we merge the current state into the test-branch? Only the passing tests?
I feel like the specs and code get split up (2 issues, 1 fork, 1 PR), should we think about consolidating?

**Comment by anthonylangsworth** [comment]
@peterwurzinger  As we said on the other thread, we're going to merge what we have. I'll do a final sanity check later tonight then push the button

**Comment by peterwurzinger** [comment]
@anthonylangsworth Yeah I missed that, sorry. But what will be the target of the merge? This repo or a pull request to dotnet/corefx?

Edit: Can someone remove my permissions to close pull request and/or issues? -.-
